Sleeping in Class solution (#3302)
cpinitiative/usaco-guide: solutions/usaco-1203.mdx

In order to test if a value of $\texttt{num\_hours}$ is valid, we iterate over
the $A$ from left to right and combine values. If at some point a combined
value exceed $\texttt{num\_hours}$ then we know that $\texttt{num\_hours}$ is not
valid.  

## Implementation

**Time Complexity:**
$\mathcal{O}(N \cdot (\# \, \text{divisors of } \texttt{sum}))$

<LanguageSection>
<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

int N;
vector<int> A;

// checks if num_hours leads to a valid answer
bool check(int num_hours) {
	int currsum = 0; // keeps track of the current sum
	// iterate over the vector
	for (int i = 0; i < N; i++) {
		currsum += A[i];
		if (currsum > num_hours)
			return false; // currsum cannot exceed x
		else if (currsum == num_hours) {
			currsum = 0; // reset currsum
		}
	}
	return true;
}

int main() {
	int T;
	cin >> T;
	while (T--) {
		// reading inputs
		cin >> N;
		A = vector<int>(N);
		int sum = 0; // sum of all values in the vector
		for (int i = 0; i < N; i++) {
			cin >> A[i];
			sum += A[i];
		}

		// num_hours is the # of hours of sleep every class AFTER modification
		// iterate over num_hours from small to large

> 		// because the smallest possible num_hours is optimal

Reviewer: ```cpp
/* 
 * multiline comments should be
 * formatted like this
 */
```

---
id: usaco-1203

> source: USACO bronze 2022 February

Reviewer: ```suggestion
source: USACO Bronze 2022 February
```

> source: USACO bronze 2022 February

Reviewer: ```suggestion
source: USACO Bronze 2022 February
```

title: Sleeping in Class
author: Yiming Huang
---

[Official Analysis (C++)](http://usaco.org/current/data/sol_prob1_bronze_feb22.html)

## Explanation

After Elsie has made all the modifications to the log array $A$, the log will
show that Bessie sleeps the same number of hours every day. Let's call that
number $\texttt{num\_hours}$.

Since the total number of hours won't change, we can call that number
$\texttt{sum}$.

Therefore, the length of the modified log will be
$\texttt{sum}/\texttt{num\_hours}$. Since every modification reduces the length
of the log by 1, the number of modifications will be
$N-\texttt{sum}/\texttt{num\_hours}$.

> $N$ and $\texttt{sum}$ are constant, so in order to minimize
> $N-\texttt{sum}/\texttt{num\_hours}$, we need to minimize $\texttt{num\_hours}$
> too.

Reviewer 2: ```suggestion
$N$ and $\texttt{sum}$ are constant, so, to minimize $N-\texttt{sum}/\texttt{num\_hours}$, 
we need to minimize $\texttt{num\_hours}$ too.
```

In order to do that, we iterate over possible values of $\texttt{num\_hours}$ and
checks if it's valid. If it is, then we print out the answer
$N-\texttt{sum}/\texttt{num\_hours}$ and return the program.

> In order to test if a value of $\texttt{num\_hours}$ is valid, we iterate over
> the $A$ from left to right and combine values. If at some point a combined
> value exceed $\texttt{num\_hours}$ then we know that $\texttt{num\_hours}$ is not
> valid.  

Reviewer 2: ```suggestion
We iterate over $A$ from left to right and combine values to test if a value 
of $\texttt{num\_hours}$ is valid. If at some point a combined value exceeds 
$\texttt{num\_hours}$ then we know that $\texttt{num\_hours}$ is not
valid.  
```

## Implementation

**Time Complexity:**
$\mathcal{O}(N \cdot (\# \, \text{divisors of } \texttt{sum}))$

<LanguageSection>
<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

int N;
vector<int> A;

/** @return if num_hours leads to a valid answer */
bool check(int num_hours) {
	int curr_sum = 0;  // keeps track of the current sum
	// iterate over the vector
	for (int i = 0; i < N; i++) {
		curr_sum += A[i];
		if (curr_sum > num_hours) {
			return false;  // currsum cannot exceed x
		} else if (curr_sum == num_hours) {
			curr_sum = 0;  // reset currsum
		}
	}
	return true;
}

int main() {
	int T;
	cin >> T;
	for (int t = 0; t < T; t++) {
		cin >> N;
		A = vector<int>(N);

> 		int sum = 0;  // sum of all values in the vector

Reviewer 3: ```suggestion
		int sum = 0;  
```
a bit extraneous comment

		for (int i = 0; i < N; i++) {
			cin >> A[i];
			sum += A[i];
		}

		/*
		 * num_hours is the # of hours of sleep every class AFTER modification
		 * iterate over num_hours from small to large
		 * because the smallest possible num_hours is optimal
		 */
		for (int num_hours = 0; num_hours <= sum; num_hours++) {
			if (num_hours != 0 && sum % num_hours != 0) {
				// sum must be divisible by num_hours
				continue;
			}
			if (check(num_hours)) {

> 				// check(num_hours) checks if num_hours is a valid answer

Reviewer 3: ```suggestion
```

$\texttt{sum}$.

Therefore, the length of the modified log will be
$\texttt{sum}/\texttt{num\_hours}$. Since every modification reduces the length
of the log by 1, the number of modifications will be
$N-\texttt{sum}/\texttt{num\_hours}$.

$N$ and $\texttt{sum}$ are constant, so in order to minimize
$N-\texttt{sum}/\texttt{num\_hours}$, we need to minimize $\texttt{num\_hours}$
too.

In order to do that, we iterate over possible values of $\texttt{num\_hours}$ and
checks if it's valid. If it is, then we print out the answer
$N-\texttt{sum}/\texttt{num\_hours}$ and return the program.

In order to test if a value of $\texttt{num\_hours}$ is valid, we iterate over
the $A$ from left to right and combine values. If at some point a combined
value exceed $\texttt{num\_hours}$ then we know that $\texttt{num\_hours}$ is not
valid.  

## Implementation

**Time Complexity:**
$\mathcal{O}(N \cdot (\# \, \text{divisors of } \texttt{sum}))$

<LanguageSection>
<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

int N;
vector<int> A;

/** @return if num_hours leads to a valid answer */
bool check(int num_hours) {

> 	int curr_sum = 0; // keeps track of the current sum

Reviewer: two spaces after line of code before beginning an inline comment

	// iterate over the vector
	for (int i = 0; i < N; i++) {
		curr_sum += A[i];
		if (curr_sum > num_hours) {
			return false; // currsum cannot exceed x
		} else if (curr_sum == num_hours) {
			curr_sum = 0; // reset currsum
		}
	}
	return true;
}

int main() {
	int T;
	cin >> T;
	for (int t = 0; t < T; t++) {
		cin >> N;
		A = vector<int>(N);
		int sum = 0; // sum of all values in the vector
		for (int i = 0; i < N; i++) {
			cin >> A[i];
			sum += A[i];
		}

		/*

> 		 *num_hours is the # of hours of sleep every class AFTER modification

Reviewer: space after asterisk

Since the total number of hours won't change, we can call that number $\texttt{sum}$.

Therefore, the length of the modified log will be $\texttt{sum}/\texttt{numHours}$. Since every modification reduces the length of the log by 1, the number of modifications will be $N-\texttt{sum}/\texttt{numHours}$.

$N$ and $\texttt{sum}$ are constant, so in order to minimize $N-\texttt{sum}/\texttt{numHours}$, we need to minimize $\texttt{numHours}$ too.

In order to do that, we iterate over possible values of $\texttt{numHours}$ and checks if it's valid. If it is, then we print out the answer $N-\texttt{sum}/\texttt{numHours}$ and return the program.

In order to test if a value of $\texttt{numHours}$ is valid, we iterate over the $A[]$ from left to right and combine values. If at some point a combined value exceed $\texttt{numHours}$ then we know that $\texttt{numHours}$ is not valid.

## Implementation

**Time Complexity:** $\mathcal{O}(N \cdot (\# \, \text{divisors of } \texttt{sum}))$

<LanguageSection>
<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

int N;
vector<int> A;

> bool check(int numHours) {

Reviewer: what this function do

	int currsum = 0;  // keeps track of the current sum
	// iterate over the vector
	for (int i = 0; i < N; i++) {
		currsum += A[i];
		if (currsum > numHours) {
			return false;  // currsum cannot exceed x
		} else if (currsum == numHours) {
			currsum = 0;  // reset currsum
		}
	}
	return true;
}

void solve() {
	// reading inputs
	cin >> N;
	A = vector<int>(N);
	int sum = 0; // sum of all values in the vector
	for (int i = 0; i < N; i++)
	{
		cin >> A[i];
		sum += A[i];
	}

	// numHours is the # of hours of sleep every class AFTER modification
	// iterate over numHours from small to large because the smallest possible numHours is optimal
	for (int numHours = 0; numHours <= sum; numHours++) {
		if (numHours != 0 && sum % numHours != 0) {
			// sum must be divisible by numHours
			continue;
		}

> 		if (check(numHours)) {

Reviewer: use snake_case for c++

---
id: usaco-1203
source: USACO bronze 2022 February
title: Sleeping in Class
author: Yiming Huang
---

[Official Analysis](http://usaco.org/current/data/sol_prob1_bronze_feb22.html)

## Explanation

> After Elsie has made all the modifications to the log array $A[]$, the log will show that Bessie sleeps the same number of hours every day. Let's call that number $\texttt{numHours}$. 

Reviewer: just say A, not A[]

$N$ and $\texttt{sum}$ are constant, so in order to minimize $N-\texttt{sum}/\texttt{numHours}$, we need to minimize $\texttt{numHours}$ too.

In order to do that, we iterate over possible values of $\texttt{numHours}$ and checks if it's valid. If it is, then we print out the answer $N-\texttt{sum}/\texttt{numHours}$ and return the program.

In order to test if a value of $\texttt{numHours}$ is valid, we iterate over the $A[]$ from left to right and combine values. If at some point a combined value exceed $\texttt{numHours}$ then we know that $\texttt{numHours}$ is not valid.

## Implementation

**Time Complexity:** $\mathcal{O}(N \cdot (\# \, \text{divisors of } \texttt{sum}))$

<LanguageSection>
<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

int N;
vector<int> A;

bool check(int numHours) {
	int currsum = 0;  // keeps track of the current sum
	// iterate over the vector
	for (int i = 0; i < N; i++) {
		currsum += A[i];
		if (currsum > numHours) {
			return false;  // currsum cannot exceed x
		}
		else if (currsum == numHours) {
			currsum = 0;  // reset currsum
		}
	}
	return true;
}

> void solve() {

Reviewer: just put this in main

	// reading inputs
	cin >> N;
	A = vector<int>(N);
	int sum = 0; // sum of all values in the vector
	for (int i = 0; i < N; i++)
	{
		cin >> A[i];
		sum += A[i];
	}

	// numHours is the # of hours of sleep every class AFTER modification

> 	// iterate over numHours from small to large because the smallest possible numHours is optimal

Reviewer: over 80 chars
